torch/functional.py

```diff
@@ -1589,3 +1590,49 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
 
 def align_tensors(*tensors):
     raise RuntimeError('`align_tensors` not yet implemented.')
+
+def result_type(*arrays_and_dtypes: Union[Tensor, torch.dtype, bool, int, float, complex]) -> torch.dtype:
+    """result_type(*arrays_and_dtypes: Union[Tensor, dtype, bool, int, float, complex]) -> dtype
+
+Returns the :class:`torch.dtype` that would result from performing an arithmetic
+operation on the provided input tensors. See type promotion :ref:`documentation <type-promotion-doc>`
+for more information on the type promotion logic.
+
+Args:
+    arrays_and_dtypes (Tensor, Number or dtype): input tensors, numbers or dtypes
+
+Example::
+
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.int), 1.0)
+    torch.float32
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.uint8), torch.tensor(1))
+    torch.uint8
+    >>> torch.result_type(torch.int32, torch.float32)
+    torch.float32
+"""
+
+    tensors = []
+    scalars: List[Union[bool, int, float, complex]] = []
+    dtypes = []
+    for x in arrays_and_dtypes:
+        if isinstance(x, (bool, int, float, complex)):
+            scalars.append(x)
+        elif isinstance(x, torch.dtype):
+            dtypes.append(x)
+        else:
+            tensors.append(x)
+    if tensors:
+        dtypes.append(_VF.result_type(tensors))  # type: ignore[attr-defined]
+    if scalars:
+        scalar_dtype = _VF._result_type_scalars(scalars)  # type: ignore[arg-type]
+        if dtypes:
+            tensor_dtype = _VF._result_type_dtypes(dtypes)
+            return _VF.result_type(_VF.tensor([], dtype=tensor_dtype),  # type: ignore[attr-defined]
+                                   _VF.tensor(0, dtype=scalar_dtype).item())
```

Comment:
Is this making use of the fact that we treat scalar tensors with lower priority than tensors for type promotion? We probably need to change this logic if we are going to follow NumPy and the array api standard and treat scalar tensors and tensors the same.

Reply:
Yes but this function correctly implements PyTorch's current logic, so it should not change